### PR TITLE
🚀 Add MVTec LOCO dataset and sPRO metric

### PR DESCRIPTION
**Pull request by djdameln**
## 📝 Description

This PR is based on the closed PR #1686 , which adds the implementation of MVTec LOCO AD dataset [[Paper](https://link.springer.com/article/10.1007/s11263-022-01578-9#Fn1)] along with the sPRO metric.

🛠️ Fixes #574 #471 #1341 

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

File: src/anomalib/cli/cli.py [hunk 1]
```diff
@@ -144,6 +144,12 @@ def add_arguments_to_parser(self, parser: ArgumentParser) -> None:
         parser.add_function_arguments(get_normalization_callback, "normalization")
         parser.add_argument("--task", type=TaskType | str, default=TaskType.SEGMENTATION)
         parser.add_argument("--metrics.image", type=list[str] | str | None, default=["F1Score", "AUROC"])
+        parser.add_argument(
+            "--metrics.pixel",
+            type=list[str] | str | dict[str, dict[str, Any]] | None,
+            default=None,
+            required=False,
+        )
```
Comment (samet-akcay) [inline src/anomalib/cli/cli.py:152]: Is this line different than Line 153?

Comment (samet-akcay) [reply src/anomalib/cli/cli.py:152]: Looks like the only difference is the type

File: src/anomalib/data/base/datamodule.py [hunk 2]
```diff
@@ -28,7 +28,9 @@
 def collate_fn(batch: list) -> dict[str, Any]:
     """Collate bounding boxes as lists.
 
-    Bounding boxes are collated as a list of tensors, while the default collate function is used for all other entries.
+    Bounding boxes and `masks` (not `mask`) are collated as a list of tensors. If `masks` is exist,
```
Comment (samet-akcay) [inline src/anomalib/data/base/datamodule.py:31]:
```suggestion
    Bounding boxes and `masks` (not `mask`) are collated as a list of tensors. If `masks` exists,
```

File: src/anomalib/data/base/datamodule.py [hunk 3]
```diff
@@ -42,6 +44,10 @@ def collate_fn(batch: list) -> dict[str, Any]:
         if "boxes" in elem:
             # collate boxes as list
             out_dict["boxes"] = [item.pop("boxes") for item in batch]
+        if "masks" in elem:
```
Comment (samet-akcay) [inline src/anomalib/data/base/datamodule.py:47]: Minor comment, but would it be an idea to rename `elem` to `sample`

File: src/anomalib/metrics/spro.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
+"""Implementation of SPRO metric based on TorchMetrics."""
+
+# Copyright (C) 2022 Intel Corporation
```
Comment (samet-akcay) [inline src/anomalib/metrics/spro.py:3]:
```suggestion
# Copyright (C) 2024 Intel Corporation
```

File: src/configs/data/mvtec_loco.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  category: breakfast_box
+  image_size: [256, 256]
+  center_crop: null
+  normalization: imagenet
+  train_batch_size: 32
+  eval_batch_size: 32
+  num_workers: 8
+  task: SEGMENTATION
+  transform_config_train: null
+  transform_config_eval: null
```
Comment (samet-akcay) [inline src/configs/data/mvtec_loco.yaml:13]: Is this still relevant? We might need update this config file

File: tests/unit/data/image/test_mvtec_loco.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+"""Unit Tests - MVTecLoco Datamodule."""
+
+# Copyright (C) 2023 Intel Corporation
```
Comment (samet-akcay) [inline tests/unit/data/image/test_mvtec_loco.py:3]:
```suggestion
# Copyright (C) 2024 Intel Corporation
```

File: tests/unit/metrics/test_spro.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
+"""Test SPRO metric."""
+
+# Copyright (C) 2023-2024 Intel Corporation
```
Comment (samet-akcay) [inline tests/unit/metrics/test_spro.py:3]:
```suggestion
# Copyright (C) 2024 Intel Corporation
```

File: src/anomalib/data/image/mvtec_loco.py [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+    MVTec LOCO AD dataset is released under the Creative Commons
+    Attribution-NonCommercial-ShareAlike 4.0 International License
+    (CC BY-NC-SA 4.0)(https://creativecommons.org/licenses/by-nc-sa/4.0/).
+
+References:
+    - Paul Bergmann, Kilian Batzner, Michael Fauser, David Sattlegger, and Carsten Steger:
+      Beyond Dents and Scratches: Logical Constraints in Unsupervised Anomaly Detection and Localization;
+      in: International Journal of Computer Vision (IJCV) 130, 947-969, 2022, DOI: 10.1007/s11263-022-01578-9
+"""
+
```
Comment (samet-akcay) [inline src/anomalib/data/image/mvtec_loco.py:18]:
```suggestion

# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

```

File: src/anomalib/data/image/mvtec_loco.py [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+logger = logging.getLogger(__name__)
+
+
+IMG_EXTENSIONS = (".png", ".PNG")
+
+DOWNLOAD_INFO = DownloadInfo(
+    name="mvtec_loco",
+    url="https://www.mydrive.ch/shares/48237/1b9106ccdfbb09a0c414bd49fe44a14a/download/430647091-1646842701"
+    "/mvtec_loco_anomaly_detection.tar.xz",
```
Comment (samet-akcay) [inline src/anomalib/data/image/mvtec_loco.py:51]: I can't seem download the dataset from this link?

File: src/anomalib/cli/cli.py [hunk 10]
```diff
@@ -144,6 +144,12 @@ def add_arguments_to_parser(self, parser: ArgumentParser) -> None:
         parser.add_function_arguments(get_normalization_callback, "normalization")
         parser.add_argument("--task", type=TaskType | str, default=TaskType.SEGMENTATION)
         parser.add_argument("--metrics.image", type=list[str] | str | None, default=["F1Score", "AUROC"])
```
Comment (ashwinvaidya17) [inline src/anomalib/cli/cli.py:146]: We should change the type for metrics.image as well

File: src/anomalib/data/image/mvtec_loco.py [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+    hashsum="d40f092ac6f88433f609583c4a05f56f",
```
Comment (samet-akcay) [inline src/anomalib/data/image/mvtec_loco.py:52]:
```suggestion
    hashsum="9e7c84dba550fd2e59d8e9e231c929c45ba737b6b6a6d3814100f54d63aae687",
```

File: src/anomalib/cli/cli.py (hunk 10 above)
Comment (ashwinvaidya17) [inline src/anomalib/cli/cli.py:146]:
```suggestion
        parser.add_argument(
             "--metrics.image",
             type=list[str] | str | dict[str, dict[str, Any]] | None,
             default=["F1Score", "AUROC"],
         )
```

File: src/anomalib/data/image/mvtec_loco.py [hunk 12]
```diff
@@ -0,0 +1,466 @@
...
+            mask = (
+                Mask(torch.zeros(image.shape[-2:])).to(torch.uint8)
+                if label_index == LabelName.NORMAL
+                else read_mask(mask_path, as_tensor=True)
+            )
+            item["image"], item["mask"] = self.transform(image, mask) if self.transform else (image, mask)
+
+            item["mask_path"] = mask_path
+            # List of masks with the original size for saturation based metrics calculation
+            item["masks"] = mask
```
Comment (samet-akcay) [inline src/anomalib/data/image/mvtec_loco.py:278]: Can we use a more descriptive name here? Not sure if this is clear from the first glance

File: src/anomalib/metrics/spro.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+            >>> spro.compute()
+            tensor(0.5433)
+        """
+        if self.total == 0:  # only background/normal images
+            return torch.Tensor([1.0])
+        return self.score / self.total
+
+
+def spro_score(
```
Comment (samet-akcay) [inline src/anomalib/metrics/spro.py:106]:
```suggestion
def compute_spro_score(
```

File: src/anomalib/metrics/spro.py [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+                list is a tensor list of masks for the corresponding image.
+
+        Example:
+            To update the metric state for the current batch, use the ``update`` method:
+
+            >>> spro.update(preds, labels)
+        """
+        score, total = spro_score(
+            predictions,
+            masks,
```
Comment (samet-akcay) [inline src/anomalib/metrics/spro.py:85]:
```suggestion
            predictions=predictions,
            targets=masks,
```

File: src/anomalib/metrics/spro.py [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+    threshold: float = 0.5,
+    saturation_config: dict | None = None,
+) -> torch.Tensor:
+    """Calculate the SPRO score for a batch of predictions.
+
+    Args:
+        predictions (torch.Tensor): Predicted anomaly masks.
+        targets: (list[torch.Tensor]): Ground truth anomaly masks with original height and width. Each element in the
+            list is a tensor list of masks for the corresponding image.
+        threshold (float): When predictions are passed as float, the threshold is used to binarize the predictions.
```
Comment (samet-akcay) [inline src/anomalib/metrics/spro.py:118]:
```suggestion
        threshold (float): When predictions are passed as float, the threshold is used to binarize the predictions.
                Defaults: ``0.5``.
```

File: src/anomalib/data/base/datamodule.py (hunk 3 above)
Comment (djdameln) [reply src/anomalib/data/base/datamodule.py:47]: We use `elem` to match the naming in the default `collate` function from torch. Let's leave it like this, it's outside of the scope of this PR anyway.

Comment (samet-akcay) [reply src/anomalib/data/base/datamodule.py:47]: alright, let's leave it as it not-related to this PR, but poor choice by torch

File: src/anomalib/data/image/mvtec_loco.py (hunk 11 above)
Comment (djdameln) [reply src/anomalib/data/image/mvtec_loco.py]: already fixed

File: src/anomalib/metrics/spro.py (hunk 13 above)
Comment (djdameln) [reply src/anomalib/metrics/spro.py:106]: Not sure about this one. I think `spro_score` is in line with TorchMetrics naming convention, where the functional metrics carry the name of the metric (so without `compute`).

File: src/anomalib/callbacks/metrics.py [hunk 16]
```diff
@@ -182,16 +181,37 @@ def _update_metrics(
         image_metric.update(output["pred_scores"], output["label"].int())
         if "mask" in output and "anomaly_maps" in output:
             pixel_metric.to(self.device)
-            pixel_metric.update(torch.squeeze(output["anomaly_maps"]), torch.squeeze(output["mask"].int()))
+            if "semantic_mask" in output:
+                self._update_pixel_metrics(pixel_metric, output)
+            else:
+                pixel_metric.update(torch.squeeze(output["anomaly_maps"]), torch.squeeze(output["mask"].int()))
```
Comment (djdameln) [inline src/anomalib/callbacks/metrics.py:187]: I'm not happy with this part of the design, as I feel that the metric callback should not contain logic related to specific metrics.

Comment (samet-akcay) [reply src/anomalib/callbacks/metrics.py:187]: agree, this part should be more generic and does not contain any specific implementation.

Comment (djdameln) [reply src/anomalib/callbacks/metrics.py:187]: It also goes against the intended use of `MetricCollection`, which is to group metrics together that operate over the same target and predictions. The sPRO metric operates over semantic masks instead of binary masks, so technically should be in a different metric collection compared to the other pixel metrics.

File: src/anomalib/callbacks/metrics.py [hunk 17]
```diff
@@ -85,9 +84,22 @@ def setup(
...
+        semantic_pixel_metric_names: list[str] | dict[str, dict[str, Any]]
+        if "SPRO" in pixel_metric_names:
+            if isinstance(pixel_metric_names, list):
+                pixel_metric_names.remove("SPRO")
+                semantic_pixel_metric_names = ["SPRO"]
+            elif isinstance(pixel_metric_names, dict):
+                spro_metric = pixel_metric_names.pop("SPRO")
+                semantic_pixel_metric_names = {"SPRO": spro_metric}
+            else:
+                logger.warning("Unexpected type for pixel_metric_names: %s", type(pixel_metric_names))
```
Comment (samet-akcay) [inline src/anomalib/callbacks/metrics.py:101]: Would it be an idea to add a comment here explaining why we are doing this. Would be good to have it for future reference, and for external readers.

Comment (djdameln) [reply src/anomalib/callbacks/metrics.py]: done

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for polishing this. I have one comment.

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks

**Review by samet-akcay — APPROVED** [review]
